### PR TITLE
ROSAENG-8274 | feat(hcp): add spot instance support for machine pools

### PR DESCRIPTION
**Pull request by amandahla**
## PR Summary

Add EC2 Spot Instance support to ROSA HCP: `spot_termination_queue_url` on `rhcs_cluster_rosa_hcp` and `use_spot_instances`/`max_spot_price` on `rhcs_hcp_machine_pool`, with plan-time validators, region-match enforcement, and subsystem test coverage. Also adds a Create/Update validation parity rule to developer-docs.

## Detailed Description of the Issue

ROSA HCP supports running worker nodes on EC2 Spot Instances to reduce compute costs. The provider lacked attributes to configure spot instance behavior at both cluster and node-pool level. This change exposes two OCM API capabilities:

- **Cluster-level** — `spot_termination_queue_url` maps to `aws.termination_handler_queue_url` in the OCM API, allowing users to specify an SQS queue for spot termination notices. Enables Enhanced mode (AWS Node Termination Handler) vs Simple mode (reactive MachineHealthCheck). Validated at plan time for SQS URL format and at apply time for region match against `cloud_region`. Region validation is enforced on both Create and Update paths.
- **Node pool-level** — `use_spot_instances` (bool) and `max_spot_price` (float64) on `aws_node_pool`, mirroring the classic machine pool schema. Maps to `aws_node_pool.spot_market_options` in the OCM API. Omitting `max_spot_price` uses on-demand pricing. Both attributes are immutable after creation — enforced at plan time via `ImmutableBool`/`ImmutableFloat64` plan modifiers (following the `ImmutableString` pattern used by `capacity_reservation_id`), with a runtime defense-in-depth check in `validateNoImmutableAttChange`. Spot and capacity reservation are mutually exclusive. A warning is emitted when creating a spot pool on a cluster without `spot_termination_queue_url`.
- **Developer docs** — Adds a MUST rule to `resource-create.md`, `resource-update.md`, and `resource-schema.md` requiring in-method validations (cross-field, region matching, mutual exclusion) to be mirrored between Create and Update. Schema-level validators run automatically on all operations, but in-method checks do not — this gap caused the region mismatch bug found during review.

## Related Issues and PRs

- Jira: [ROSAENG-8274](https://issues.redhat.com/browse/ROSAENG-8274)
- Fixes: N/A
- Related PR(s): [openshift/rosa#3457](https://github.com/openshift/rosa/pull/3457) (CLI implementation)
- Related design/docs: `aws-spot-instances-rosa-hcp-client-contract.md`

## Type of Change

- [x] feat - adds a new user-facing capability.
- [ ] fix - resolves an incorrect behavior or bug.
- [x] docs - updates documentation only.
- [ ] style - formatting or naming changes with no logic impact.
- [ ] refactor - code restructuring with no behavior change.
- [ ] test - adds or updates tests only.
- [ ] chore - maintenance work (tooling, housekeeping, non-product code).
- [ ] build - changes build system, packaging, or dependencies for build output.
- [ ] ci - changes CI pipelines, jobs, or automation workflows.
- [ ] perf - improves performance without changing intended behavior.

## Previous Behavior

No Terraform attributes existed to configure EC2 Spot Instances on ROSA HCP clusters or node pools. Developer docs did not call out the Create/Update validation parity requirement.

## Behavior After This Change

- `rhcs_cluster_rosa_hcp` resource and data source expose an optional `spot_termination_queue_url` string attribute (validated as an HTTPS SQS URL with correct region).
- `rhcs_hcp_machine_pool` resource and data source expose `use_spot_instances` (bool) and `max_spot_price` (float64) under `aws_node_pool`, matching the classic `rhcs_machine_pool` schema.
- All spot-related attributes are `Optional` and round-trip cleanly through the OCM API.
- Update path supports setting, changing, and clearing `spot_termination_queue_url` (following the `audit_log_arn` pattern). Region mismatch is rejected on both Create and Update.
- `use_spot_instances` and `max_spot_price` are immutable after creation — enforced at **plan time** via `ImmutableBool`/`ImmutableFloat64` plan modifiers (same pattern as `ImmutableString` on `capacity_reservation_id`), with a runtime defense-in-depth check via `validateOptionalOnlyImmutable` in the Update path. This catches both value changes (e.g. `true → false`) and attribute removal (e.g. `true → <null>`).
- Spot and capacity reservation are mutually exclusive (rejected at Create time).
- A warning is emitted when creating a spot node pool on a cluster without `spot_termination_queue_url`.
- Developer docs (`resource-create.md`, `resource-update.md`, `resource-schema.md`) now require in-method validations to be mirrored between Create and Update.

## ROSA CLI Parity

Compared against [openshift/rosa#3457](https://github.com/openshift/rosa/pull/3457). Core validation and business logic matches. Intentional differences:

| Area | ROSA CLI | Terraform Provider | Rationale |
|---|---|---|---|
| **SQS URL validation** | Lightweight HTTPS-only checks (`ValidateHTTPSQueueURL`) | Stricter: validates SQS host pattern, path format (`/<account-id>/<queue-name>`), and region match against `cloud_region` | Provider runs at plan time without interactive correction — stricter validation catches mistakes earlier |
| **Local zone spot rejection** | Checks `IsLocalAvailabilityZone()` via AWS API, rejects spot | Not checked | Requires AWS API call not available to the RHCS provider; OCM backend enforces this constraint |
| **Spot without termination handler warning** | Warns via `Reporter.Warnf` on `rosa create machinepool` | Emits `Diagnostics.AddWarning` during `terraform apply` | Added to match CLI guidance |
| **Spot immutability on edit** | Rejects add/remove/change max_price | Same — `ImmutableBool`/`ImmutableFloat64` plan modifiers on `use_spot_instances` and `max_spot_price`, plus runtime `validateOptionalOnlyImmutable` | Full parity; plan-time enforcement gives better UX than apply-time rejection |
| **Spot + capacity reservation** | Rejects at create | Same — rejects at create | Full parity |
| **Cluster termination queue URL update** | Allows set/change/clear | Same — allows set/change/clear | Full parity |
| **max_price semantics** | `"on-demand"` string default | `null`/omitted = on-demand | Different UX, equivalent OCM API behavior |
| **Node pool spot schema** | `--use-spot-instances` (bool) + `--spot-max-price` (string) | `use_spot_instances` (bool) + `max_spot_price` (float64) inside `aws_node_pool` — mirrors classic `rhcs_machine_pool` | We chose flat attributes over the nested `spot_market_options` block proposed in the [enhancement](https://github.com/openshift-online/rosa-enhancements/pull/59). The flat design cleanly represents the three spot states: (1) no spot attrs = regular on-demand, (2) `use_spot_instances = true` without `max_spot_price` = spot capped at on-demand price, (3) both set = spot with explicit price cap. A nested block cannot distinguish cases 1 and 2 without an empty block (`spot_market_options = {}`), which is unintuitive. The attributes remain inside `aws_node_pool` as the enhancement intended — only the nesting level changed. Feature, validation, and business logic are the same. |

### Preconditions

- `TF_ACC=1` and valid OCM credentials for subsystem tests (or use mock server).
- Go 1.25+ installed.

### Test Steps

1. Run cluster spot termination queue URL subsystem tests:
   ```bash
   TF_ACC=1 go test ./subsystem/hcp/... -v -run TestAcc_ClusterRosaHcp_SpotTermination -timeout 30m
   ```
2. Run machine pool spot market options subsystem tests:
   ```bash
   TF_ACC=1 go test ./subsystem/hcp/... -v -run TestAcc_MachinePool_Spot -timeout 30m
   ```
3. Run validator unit tests:
   ```bash
   go test ./provider/common/attrvalidators/... -v
   ```

### Expected Results

All tests pass. Spot attributes are correctly sent to the OCM API on create, read back into state on refresh, and produce no spurious diffs on subsequent plans. Region mismatches are rejected on both create and update.

## Proof of the Fix

- Logs/CLI output: subsystem and unit test results from CI.

### `main.tf`

  ```hcl
  resource "rhcs_cluster_rosa_hcp" "cluster" {
    name           = "test-cluster"
    cloud_region   = var.aws_region
    aws_account_id = data.aws_caller_identity.current.account_id
    # ...

    spot_termination_queue_url = "https://sqs.us-west-2.amazonaws.com/<aws-account-id>/<queue-name>"

    # ...
  }

  resource "rhcs_hcp_machine_pool" "spot_pool" {
    cluster = rhcs_cluster_rosa_hcp.cluster.id
    name    = "spot-pool"

    aws_node_pool = {
      instance_type      = "m5.xlarge"
      use_spot_instances = true
      max_spot_price     = 2.0
    }

    subnet_id   = var.private_subnet_id
    replicas    = 2
    auto_repair = true
    # ...
  }

  resource "rhcs_hcp_machine_pool" "spot_ondemand" {
    cluster = rhcs_cluster_rosa_hcp.cluster.id
    name    = "spot-od"

    aws_node_pool = {
      instance_type      = "m5.xlarge"
      use_spot_instances = true
    }

    subnet_id   = var.private_subnet_id
    replicas    = 2
    auto_repair = true
    # ...
  }

  resource "rhcs_hcp_machine_pool" "regular_pool" {
    cluster = rhcs_cluster_rosa_hcp.cluster.id
    name    = "regular"

    aws_node_pool = {
      instance_type = "m5.xlarge"
    }

    subnet_id   = var.private_subnet_id
    replicas    = 2
    auto_repair = true
    # ...
  }
  ```

  ### Test Results

  #### 1. Region validation works

  Applying a queue URL in a different region than the cluster is correctly rejected:

  ```
  ╷
  │ Error: Can't update cluster
  │
  │   with rhcs_cluster_rosa_hcp.cluster,
  │   on main.tf line 45, in resource "rhcs_cluster_rosa_hcp" "cluster":
  │
  │ Spot termination queue URL region 'us-east-2' does not match cluster region
  │ 'us-west-2'. The SQS queue must be in the same region as the cluster.
  ╵
  ```

  #### 2. Cluster queue URL change works

  The `spot_termination_queue_url` attribute is successfully updated on the cluster.

  #### 3. Cluster description shows the queue URL

  ```
  $ rosa describe cluster --cluster=<cluster-id>
  Spot Termination Queue URL: https://sqs.us-west-2.amazonaws.com/<aws-account-id>/<queue-name>
  ```

  #### 4. Machine pools are created as expected

  ```
  $ rosa list machinepool --cluster=<cluster-id>
  ID         AUTOSCALING  REPLICAS  INSTANCE TYPE  SPOT INSTANCES   DISK SIZE  VERSION  AUTOREPAIR
  regular    No           2/2       m5.xlarge      No               300 GiB    4.22.8   Yes
  spot-od    No           2/2       m5.xlarge      Yes (on-demand)  300 GiB    4.22.8   Yes
  spot-pool  No           2/2       m5.xlarge      Yes (max $1)     300 GiB    4.22.8   Yes
  workers    No           2/2       m5.xlarge      No               300 GiB    4.22.8   Yes
  ```

  #### 5. Changing spot price is blocked

  ```
  ╷
  │ Error: Attribute value cannot be changed
  │
  │   with rhcs_hcp_machine_pool.spot_pool,
  │
  │ Attribute aws_node_pool.max_spot_price, cannot be changed from 1.000000
  │ to 2.000000
  ╵
  ```

  #### 6. Disabling spot instances is blocked

  Removing `use_spot_instances` or setting it to `false` are both rejected:

  ```
  ╷
  │ Error: Attribute value cannot be changed
  │
  │   with rhcs_hcp_machine_pool.spot_ondemand,
  │
  │ Attribute aws_node_pool.use_spot_instances, cannot be changed from true
  │ to <null>
  ╵

  ╷
  │ Error: Attribute value cannot be changed
  │
  │   with rhcs_hcp_machine_pool.spot_ondemand,
  │
  │ Attribute aws_node_pool.use_spot_instances, cannot be changed from true
  │ to false
  ╵
  ```

## Breaking Changes

- [x] No breaking changes
- [ ] Yes, this PR introduces a breaking change (describe impact and migration plan below; see [breaking-change guidance](https://github.com/terraform-redhat/terraform-provider-rhcs/blob/main/developer-docs/breaking-changes.md))

## Developer Verification Checklist
- [x] Commit subject/title follows `[JIRA-TICKET] | [TYPE][(scope)][!]: <MESSAGE>`.
- [x] PR description clearly explains both **what** changed and **why**.
- [x] Relevant Jira/GitHub issues and related PRs are linked.
- [x] `make install-hooks` has been run in this clone.
- [x] `make pre-push-checks` passes.
- [x] Documentation was added/updated where appropriate (see [docs and examples](https://github.com/terraform-redhat/terraform-provider-rhcs/blob/main/developer-docs/docs-and-examples.md)).
- [ ] Any risk, limitation, or follow-up work is documented.
- [x] **Classic and HCP:** If this PR touches both Classic and HCP paths, I called out parity or intentional divergence (see [architecture](https://github.com/terraform-redhat/terraform-provider-rhcs/blob/main/developer-docs/architecture.md)).
- [ ] **Auth / secrets / sensitive:** Changes involving credentials, tokens, Sensitive attributes, or request/response logging follow [security](https://github.com/terraform-redhat/terraform-provider-rhcs/blob/main/developer-docs/security.md).

### Testing (check all that apply; use N/A when not relevant)
- [ ] **N/A** — no provider resource/data source or `provider/` / `internal/` logic changes.
- [x] **New or changed resource / data source** — subsystem test added or updated under `subsystem/classic/` or `subsystem/hcp/` (see [testing](https://github.com/terraform-redhat/terraform-provider-rhcs/blob/main/developer-docs/testing.md), [resource overview](https://github.com/terraform-redhat/terraform-provider-rhcs/blob/main/developer-docs/resources/resource-overview.md), and [data source overview](https://github.com/terraform-redhat/terraform-provider-rhcs/blob/main/developer-docs/datasources/datasource-overview.md)).
- [x] **New or changed validation, plan modifiers, or helpers** — unit tests in the same package (`*_test.go`), or a subsystem negative test when integration-only (not both for the same cases unless a wiring smoke test is needed).
- [x] **Schema / config validation** — unit test and/or subsystem test expecting plan/apply failure (one primary layer per rule; see [CONTRIBUTING.md](https://github.com/terraform-redhat/terraform-provider-rhcs/blob/main/CONTRIBUTING.md) and [testing](https://github.com/terraform-redhat/terraform-provider-rhcs/blob/main/developer-docs/testing.md)).
- [x] **Validators / helpers** — unit tests in the same package where applicable (review policy; no automated coverage % gate).
- [ ] `make check-subsystem-registry` passes.
- [x] I manually tested the change when behavior is user-visible.

## Summary by CodeRabbit

- **New Features**
  - Configure an optional SQS termination queue for enhanced Spot Instance termination handling on hosted clusters.
  - View the configured termination queue URL in cluster data sources.
  - Configure Spot Market Options for AWS machine pools, including an optional maximum price.
  - View Spot configuration and maximum price in machine pool data sources.

- **Bug Fixes**
  - Added validation for SQS URLs, queue regions, and valid Spot maximum prices.
  - Prevented unsupported Spot configuration changes and conflicts with capacity reservations.

- **Documentation**
  - Updated resource and data source documentation with the new settings and behavior.

### CODE REVIEW

File: provider/clusterrosa/hcp/resource.go [hunk 1]
```diff
@@ -1646,6 +1671,16 @@ func (r *ClusterRosaHcpResource) Update(ctx context.Context, request resource.Up
...
+	if plan.SpotTerminationQueueUrl.IsNull() && !state.SpotTerminationQueueUrl.IsNull() &&
+		state.SpotTerminationQueueUrl.ValueString() != "" {
+		awsBuilder.TerminationHandlerQueueUrl("")
+		changesToAws = true
+	} else if newUrl, shouldPatch := common.ShouldPatchString(
+		state.SpotTerminationQueueUrl, plan.SpotTerminationQueueUrl); shouldPatch {
+		awsBuilder.TerminationHandlerQueueUrl(newUrl)
+		changesToAws = true
+	}
+
```
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/clusterrosa/hcp/resource.go:1686]: Fixed. Added the same `ExtractRegionFromSqsUrl` region-mismatch check to the Update path (mirrors the Create path at lines 725-735). Also added a subsystem test for "update URL to wrong region". See commit 23eab909.

File: provider/common/attrvalidators/spot_price_validator.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	// Validate that the number is positive (greater than zero)
+	if price <= 0 {
+		response.Diagnostics.AddAttributeError(
+			request.Path,
+			"Invalid Spot max price",
+			fmt.Sprintf("Spot max price must be a positive decimal number, got %q", value),
+		)
+		return
```
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/common/attrvalidators/spot_price_validator.go]: Fixed. Added `math.IsNaN(price) || math.IsInf(price, 0)` check after `ParseFloat`, before the positivity check. `strconv.ParseFloat("NaN", 64)` returns `NaN, nil` and `NaN <= 0` is `false` per IEEE 754, so this was a real gap. Added test entries for `"NaN"`, `"Inf"`, and `"Infinity"`. See commit 23eab909.

File: provider/machinepool/hcp/machine_pool_resource.go [hunk 3]
```diff
@@ -801,12 +823,34 @@ func (r *HcpMachinePoolResource) doUpdate(ctx context.Context, state *HcpMachine
...
+		if plan.AWSNodePool.SpotMarketOptions != nil {
+			spotBuilder := cmv1.NewAwsNodePoolSpotMarketOptions()
+			if common.HasValue(plan.AWSNodePool.SpotMarketOptions.MaxPrice) {
+				maxPrice := plan.AWSNodePool.SpotMarketOptions.MaxPrice.ValueString()
+				if maxPrice != "" {
+					spotBuilder.MaxPrice(maxPrice)
+				}
+			}
+			awsNodePoolBuilder.SpotMarketOptions(spotBuilder)
+		}
```
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/machinepool/hcp/machine_pool_resource.go]: Acknowledged. The Create path already rejects spot + capacity reservation. On the Update path, capacity reservation fields are immutable (enforced at lines 810-824), so the only scenario is adding spot to an existing capacity-reservation pool — which the OCM API rejects server-side. Adding a client-side check here would give a better error message but isn't strictly a correctness issue. Will consider adding for UX in a follow-up if needed.

File: provider/clusterrosa/hcp/resource.go [hunk 4]
```diff
@@ -259,6 +259,17 @@ func (r *ClusterRosaHcpResource) Schema(ctx context.Context, req resource.Schema
...
+				Description: "The URL of an SQS queue for enhanced spot termination handling. " +
+					"When set, enables Enhanced mode using AWS Node Termination Handler to " +
+					"gracefully drain nodes within the 2-minute spot interruption window. " +
+					"When omitted, Simple mode is used (reactive replacement via MachineHealthCheck). " +
+					"The queue must be a standard SQS queue in the same region as the cluster.",
+				Optional: true,
+				Validators: []validator.String{
+					attrvalidators.SqsUrlValidator(),
+				},
+			},
```
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/clusterrosa/hcp/resource.go:272]: This is the same finding as the other comment on the Update path (resource.go:1683). Already fixed — added `ExtractRegionFromSqsUrl` region check to the Update path, mirroring the Create path. Also added a subsystem test for update-time region mismatch. See commit 23eab909.

File: provider/common/attrvalidators/spot_price_validator_test.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+		),
+		Entry("unknown value -> ok",
+			validator.StringRequest{
+				Path:           path.Root("spot_max_price"),
+				PathExpression: path.MatchRoot("spot_max_price"),
+				ConfigValue:    types.StringUnknown(),
+			},
+			false,
+		),
+	)
```
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/common/attrvalidators/spot_price_validator_test.go]: Intentionally not implemented. The `"on-demand"` sentinel was considered but deliberately dropped — on-demand pricing is achieved by simply omitting `max_price` (the field is Optional). This is cleaner because a sentinel string like `"on-demand"` would cause perpetual Terraform drift: the user writes `max_price = "on-demand"`, the provider skips sending it to the API, the API returns no max_price, populateState sets null, and the next plan detects `"on-demand"` != null → infinite apply cycles. The old PR description mentioned `"on-demand"` but has been updated to reflect the actual implementation.

File: provider/common/planmodifiers/immutable_bool_modifier.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	if req.PlanValue.Equal(req.StateValue) {
+		return
+	}
+
+	resp.Diagnostics.AddError(
+		common.AssertionErrorSummaryMessage,
+		fmt.Sprintf(common.AssertionErrorDetailsMessage, req.Path.String(), req.StateValue, req.PlanValue),
+	)
```
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/common/planmodifiers/immutable_bool_modifier.go:63]: Done — `IsUnknown()` early return was already added to both `PlanModifyBool` and `PlanModifyFloat64` before comparing with `Equal`. The guard prevents false immutability errors when plan values come from interpolation (e.g. `use_spot_instances = var.enable_spot`). Terraform re-validates once the value resolves.

Tests were not added for the bool/float64 modifiers specifically — the existing `immutable_string_modifier_test.go` covers the pattern and the string modifier will get the same guard in a follow-up.

File: provider/machinepool/hcp/aws_node_pool.go [hunk 7]
```diff
@@ -109,6 +112,23 @@ func AwsNodePoolResource() map[string]schema.Attribute {
 				stringplanmodifier.UseStateForUnknown(),
 			},
 		},
+		"use_spot_instances": schema.BoolAttribute{
+			Description: "Use Amazon EC2 Spot Instances. " + common.ValueCannotBeChangedStringDescription,
+			Optional:    true,
+			PlanModifiers: []planmodifier.Bool{
+				commonplanmodifiers.ImmutableBool(),
+			},
```
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/machinepool/hcp/aws_node_pool.go:125]: Evaluated — this is a real edge case but not a practical issue, so deferring for now.

**Analysis:** When a user omits `use_spot_instances` (the normal non-spot case), Read correctly sets it to `null`, which matches the config — no drift. The drift only occurs if a user explicitly writes `use_spot_instances = false`, which is a no-op (identical to omitting it). In practice nobody would do this.

**Why not fix now:** The clean fix (Optional+Computed) would be a state contract change that needs migration testing. The workaround (normalize in Read) adds complexity for a scenario with near-zero real-world occurrence. Both options carry more risk than the bug itself.

**What we did instead:** Added a rule to `developer-docs/resources/resource-schema.md` documenting the `false` vs `null` footgun for Optional booleans, so future contributors avoid this pattern or handle it explicitly.

File: provider/machinepool/hcp/aws_node_pool.go [hunk 8]
```diff
@@ -109,6 +112,23 @@ func AwsNodePoolResource() map[string]schema.Attribute {
...
+		},
+		"max_spot_price": schema.Float64Attribute{
+			Description: "Max Spot price. " + common.ValueCannotBeChangedStringDescription,
+			Optional:    true,
+			Validators: []validator.Float64{
+				float64validator.AtLeast(1e-6),
+			},
+			PlanModifiers: []planmodifier.Float64{
+				commonplanmodifiers.ImmutableFloat64(),
+			},
```
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/machinepool/hcp/aws_node_pool.go:137]: Done — schema descriptions for both attributes now document the constraints:

- `use_spot_instances`: mentions `max_spot_price` relationship, capacity reservation mutual exclusion, and immutability
- `max_spot_price`: documents that it requires `use_spot_instances = true`, minimum value of 1e-6, and on-demand fallback when omitted

Docs will be regenerated with `make docs` before merge.

File: provider/machinepool/hcp/aws_node_pool.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/machinepool/hcp/aws_node_pool.go:125]: Update: we reconsidered this edge case and addressed it with an `IgnoreFalse()` plan modifier on `use_spot_instances`.

Instead of preserving `false` in state (which would require an Optional+Computed schema change and migration testing), we normalize `false` → `null` in the plan phase. If a user writes `use_spot_instances = false`, the modifier treats it as "not set" — nothing is sent to the API, nothing is stored in state, and there's no drift on refresh.

The modifier runs before `ImmutableBool()` in the plan modifier chain, so immutability checks still work correctly against the normalized value.

This avoids the state contract change you flagged while eliminating the drift scenario entirely.

**Comment by amandahla** [comment]
@coderabbitai Review the full spot instance implementation in this PR. Focus on:

  1. **ROSA CLI parity** — Do the spot instance attributes (`use_spot_instances`, `max_spot_price`, `spot_termination_queue_url`) match the behavior and constraints of `rosa create machinepool --use-spot-instances --spot-max-price` and `rosa
  edit cluster --spot-termination-queue-url`? Any gaps or divergences?
  2. **Terraform UX** — Schema descriptions, error messages, plan modifier behavior, validation flow. Is the practitioner experience clear and consistent? Any remaining drift scenarios?
  3. **Immutability enforcement** — The dual approach (plan modifiers + manual validation in Update). Is the coverage complete?
  4. **SQS URL validator** — Security, correctness, FIPS endpoint support, region extraction.
  5. **Test coverage** — Are the subsystem tests comprehensive for positive, negative, and edge cases?

  If the implementation is solid, approve. No need to re-flag items already addressed in previous review threads.

**Comment by amandahla** [comment]
@coderabbitai confirm that findings in the 'ci/prow/security' check are not result of this PR.
Same for the ci/prow/govulncheck

**Comment by amandahla** [comment]
/override ci/prow/govulncheck
/override ci/prow/security
See https://github.com/terraform-redhat/terraform-provider-rhcs/pull/1300#issuecomment-5319405616

File: provider/clusterrosa/hcp/resource.go [hunk 9]
```diff
@@ -259,6 +259,17 @@ func (r *ClusterRosaHcpResource) Schema(ctx context.Context, req resource.Schema
 					"to send audit logs to a CloudWatch Logs log group. To disable audit log forwarding, provide an empty string.",
 				Optional: true,
 			},
+			"spot_termination_queue_url": schema.StringAttribute{
+				Description: "The URL of an SQS queue for enhanced spot termination handling. " +
+					"When set, enables Enhanced mode using AWS Node Termination Handler to " +
+					"gracefully drain nodes within the 2-minute spot interruption window. " +
+					"When omitted, Simple mode is used (reactive replacement via MachineHealthCheck). " +
+					"The queue must be a standard SQS queue in the same region as the cluster.",
```
Comment (olucasfreitas) [inline provider/clusterrosa/hcp/resource.go]: The implementation here appears to allow FIFO queue URLs, and the HyperShift API docs say both standard and FIFO are supported. I’d soften the description from 'standard SQS queue' to 'SQS queue' or explicitly mention FIFO so the provider docs don’t understate what the backend accepts.

File: provider/common/planmodifiers/ignore_false_modifier.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// IgnoreFalse returns a plan modifier that normalizes an explicit false to null,
+// making false equivalent to omitting the attribute entirely.
+func IgnoreFalse() planmodifier.Bool {
```
Comment (olucasfreitas) [inline provider/common/planmodifiers/ignore_false_modifier.go:15]: These helpers are now shared Terraform contract code, not one-off resource logic. I’d still add same-package tests for `IgnoreFalse()`, `ImmutableBool()`, and `ImmutableFloat64()` here before merge, especially since the existing `ImmutableString` tests currently encode different unknown-plan behavior.

File: provider/machinepool/hcp/aws_node_pool.go [hunk 11]
```diff
@@ -109,6 +112,31 @@ func AwsNodePoolResource() map[string]schema.Attribute {
...
+				commonplanmodifiers.ImmutableBool(),
+			},
+		},
+		"max_spot_price": schema.Float64Attribute{
+			Description: "Maximum hourly price for Spot Instances in USD. Requires use_spot_instances to be true. " +
+				"Must be at least 0.000001 (1e-6). If not specified, the on-demand price is used as the maximum. " +
+				common.ValueCannotBeChangedStringDescription,
+			Optional: true,
+			Validators: []validator.Float64{
+				float64validator.AtLeast(1e-6),
```
Comment (olucasfreitas) [inline provider/machinepool/hcp/aws_node_pool.go]: The `1e-6` floor here is looser than the backend contract. AWS Spot and the HyperShift API model both require max price to be at least `$0.001`, so this will still allow invalid plans through to apply. Can we tighten the validator and docs to the real floor?

File: provider/common/planmodifiers/immutable_bool_modifier.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply provider/common/planmodifiers/immutable_bool_modifier.go:63]: The unknown-plan guard looks correct now, but I’d still add same-package tests in this PR. These modifiers now define the Terraform contract for HCP Spot settings, and the only existing immutability tests in this package currently encode different unknown-value behavior.

File: provider/machinepool/hcp/aws_node_pool.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply provider/machinepool/hcp/aws_node_pool.go:125]: I agree with using `IgnoreFalse()` instead of changing the resource state contract to Optional+Computed. The missing piece is a focused test that proves explicit `use_spot_instances = false` normalizes to null and stays stable across refresh.

[the remark this answers is not in the document]
Comment (amandahla) [reply provider/machinepool/hcp/aws_node_pool.go:125]: @olucasfreitas any action for re-opening the comment?

File: provider/clusterrosa/hcp/resource.go (hunk 9 above)
Comment (amandahla) [reply provider/clusterrosa/hcp/resource.go]: I asked on Slack about that, there are conflicting information across the docs.

File: provider/machinepool/hcp/aws_node_pool.go (hunk 11 above)
Comment (amandahla) [reply provider/machinepool/hcp/aws_node_pool.go]: Where is this information please?
"AWS Spot and the HyperShift API model both require max price to be at least $0.001"
1e-6 is 0.000001 and accepted by regex 
https://<internatl red hat>/service/uhc-clusters-service/-/blob/master/cmd/clusters-service/servecmd/apiserver/validation_helpers.go#L3532

I changed the validator anyway to accept positive values (>0) favoring ROSA CLI parity.

File: provider/common/planmodifiers/ignore_false_modifier.go (hunk 10 above)
Comment (amandahla) [reply provider/common/planmodifiers/ignore_false_modifier.go:15]: Tests added for all planmodifiers now.

Comment (amandahla) [reply provider/common/planmodifiers/ignore_false_modifier.go:15]: @coderabbitai review the tests

Comment (amandahla) [reply provider/common/planmodifiers/ignore_false_modifier.go:15]: Done — added `null→value` and `value→null` cases to both `ImmutableBool` and `ImmutableFloat64` test tables, matching the coverage in `ImmutableString`. All 32 plan modifier specs pass.

Comment (amandahla) [reply provider/common/planmodifiers/ignore_false_modifier.go:15]: @coderabbitai confirm that your review was addressed.

File: provider/common/planmodifiers/immutable_bool_modifier.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply provider/common/planmodifiers/immutable_bool_modifier.go:63]: Tests added.

File: provider/clusterrosa/hcp/resource.go (hunk 9 above)
Comment (amandahla) [reply provider/clusterrosa/hcp/resource.go]: After some discussion in the slack channel  wg-rosa26-aws-spot-market-options, since Hypershift doesnt reject it, it was decided to be accepted too. I'll change the doc and review the validation.

Comment (amandahla) [reply provider/clusterrosa/hcp/resource.go]: Done, doc/validation/tests were reviewed as well.

File: provider/common/planmodifiers/ignore_false_modifier.go (hunk 10 above)
Comment (amandahla) [reply provider/common/planmodifiers/ignore_false_modifier.go:15]: @coderabbitai I forgot to push, check again.

File: provider/machinepool/hcp/aws_node_pool.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply provider/machinepool/hcp/aws_node_pool.go:125]: No further action needed. The `IgnoreFalse()` approach is solid and the plan modifier tests added in this push cover the normalization behavior. Resolving.

File: provider/machinepool/hcp/aws_node_pool.go (hunk 11 above)
Comment (olucasfreitas) [reply provider/machinepool/hcp/aws_node_pool.go]: The $0.001 figure comes from [AWS Spot documentation](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_SpotMarketOptions.html), but the OCM CS backend regex accepts any positive value — aligning with the immediate backend contract and CLI parity makes sense here. The `PositiveFloat64()` validator is the right call.

**Comment by amandahla** [comment]
@coderabbitai check failures ci/prow/e2e-presubmits-rosa-hcp-advanced-critical-high-presubmit and ci/prow/e2e-presubmits-rosa-sts-private-critical-high-presubmit, confirm that they are not related to the PR

**Comment by amandahla** [comment]
/override ci/prow/e2e-presubmits-rosa-hcp-advanced-critical-high-presubmit
/override ci/prow/e2e-presubmits-rosa-sts-private-critical-high-presubmit
